File: examples/typescript/src/agent.ts
/**
 * Customer Data Assistant — AutomaGuard TypeScript SDK example.
 *
 * Demonstrates AutomaGuard policy enforcement on a LangChain.js agent.
 * Two prompts are included:
 *
 *   --safe    Aggregate query, no PII accessed or sent externally.
 *             All events receive `allow` or `audit` verdicts.
 *
 *   --unsafe  Adversarial prompt that attempts to exfiltrate a customer's
 *             PII via an external email. The `NoPIIExfiltration` invariant
 *             and the unapproved-domain rule block the email before it sends.
 *
 * Usage:
 *   npx tsx src/agent.ts --safe
 *   npx tsx src/agent.ts --unsafe
 */

import { ChatOpenAI } from '@langchain/openai';
import { AgentExecutor, createOpenAIFunctionsAgent } from 'langchain/agents';
import { ChatPromptTemplate, MessagesPlaceholder } from '@langchain/core/prompts';
import { AutomaGuardCallbackHandler } from 'automaguard/langchain';
import { EnforcementError } from 'automaguard';

import { tools } from './tools.js';
import { SAFE_PROMPT, UNSAFE_PROMPT } from './prompts.js';

// ── Policy setup ──────────────────────────────────────────────────────────────

const POLICY_PATH = new URL('../../customer_data_guard.aegisc', import.meta.url)
  .pathname;

const handler = new AutomaGuardCallbackHandler({
  policy: POLICY_PATH,
  onAudit: (result, toolName) => {
    console.log(
      `  [audit] ${toolName}: ${result.reason ?? 'no reason'} ` +
        `(rules: [${result.triggered_rules.join(', ')}])`
    );
  },
});

// ── Agent setup ───────────────────────────────────────────────────────────────

const llm = new ChatOpenAI({ model: 'gpt-4o', temperature: 0 });

const prompt = ChatPromptTemplate.fromMessages([
  ['system', 'You are a helpful customer support analyst assistant.'],
  ['human', '{input}'],
  new MessagesPlaceholder('agent_scratchpad'),
]);

const agent = await createOpenAIFunctionsAgent({ llm, tools, prompt });

const executor = new AgentExecutor({
  agent,
  tools,
  callbacks: [handler],
  verbose: false,
});

// ── Run ───────────────────────────────────────────────────────────────────────

const mode = process.argv.includes('--safe') ? 'safe' : 'unsafe';
const input = mode === 'safe' ? SAFE_PROMPT : UNSAFE_PROMPT;

console.log(`\n=== AutomaGuard TypeScript Example (${mode} run) ===\n`);
console.log('Prompt:', input, '\n');

try {
  const result = await executor.invoke({ input });
  console.log('\nResult:', result.output);
} catch (e) {
  if (e instanceof EnforcementError) {
    console.error('\nBLOCKED by AutomaGuard policy:');
    console.error('  Reason:', e.result.reason);
    console.error('  Verdict:', e.result.verdict);
    if (e.result.triggered_rules.length > 0) {
      console.error('  Rules triggered:', e.result.triggered_rules);
    }
    if (e.result.violations.length > 0) {
      console.error('  Invariant violations:');
      for (const v of e.result.violations) {
        console.error(`    - ${v.proof}/${v.invariant}: ${v.message}`);
      }
    }
  } else {
    console.error('\nUnexpected error:', e);
  }
  process.exit(1);
}

const status = handler.policyEngine.status();
console.log(`\n[engine] ${status.events_processed} events evaluated`);

File: examples/python/agent.py
"""Customer Data Assistant — AutomaGuard Python SDK example.

Demonstrates AutomaGuard policy enforcement on a LangChain agent.

Usage:
    python agent.py --safe    # Aggregate query, no PII accessed or sent externally
    python agent.py --unsafe  # Adversarial PII exfiltration attempt (blocked)
"""

from __future__ import annotations

import logging
import os
import sys
from pathlib import Path

from langchain.agents import AgentExecutor, create_openai_functions_agent
from langchain.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_openai import ChatOpenAI

from aegis_enforce import AutomaGuardCallbackHandler, EnforcementError, PolicyEngine

from prompts import SAFE_PROMPT, UNSAFE_PROMPT
from tools import delete_record, get_customer_profile, human_approved, query_tickets, send_email

# ── Policy setup ──────────────────────────────────────────────────────────────

POLICY_PATH = Path(__file__).parent.parent / "customer_data_guard.aegisc"

if not POLICY_PATH.exists():
    print(
        f"Policy bytecode not found at {POLICY_PATH}.\n"
        "Compile it first:\n"
        "  aegisc compile examples/customer_data_guard.aegis "
        "-o examples/customer_data_guard.aegisc",
        file=sys.stderr,
    )
    sys.exit(1)


def on_audit(result, tool_name: str = "tool") -> None:
    print(
        f"  [audit] {tool_name}: {result.reason or 'no reason'} "
        f"(rules: {result.triggered_rules})"
    )


handler = AutomaGuardCallbackHandler(
    policy=str(POLICY_PATH),
    on_deny="raise",
)

# ── Agent setup ───────────────────────────────────────────────────────────────

tools = [get_customer_profile, query_tickets, send_email, delete_record, human_approved]

llm = ChatOpenAI(model="gpt-4o", temperature=0)

prompt = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a helpful customer support analyst assistant."),
        ("human", "{input}"),
        MessagesPlaceholder(variable_name="agent_scratchpad"),
    ]
)

agent = create_openai_functions_agent(llm=llm, tools=tools, prompt=prompt)

executor = AgentExecutor(
    agent=agent,
    tools=tools,
    callbacks=[handler],
    verbose=False,
)

# ── Run ───────────────────────────────────────────────────────────────────────

mode = "safe" if "--safe" in sys.argv else "unsafe"
input_text = SAFE_PROMPT if mode == "safe" else UNSAFE_PROMPT

print(f"\n=== AutomaGuard Python Example ({mode} run) ===\n")
print(f"Prompt: {input_text}\n")

try:
    result = executor.invoke({"input": input_text})
    print(f"\nResult: {result['output']}")
except EnforcementError as e:
    print("\nBLOCKED by AutomaGuard policy:")
    print(f"  Reason:  {e.result.reason}")
    print(f"  Verdict: {e.result.verdict}")
    if e.result.triggered_rules:
        print(f"  Rules triggered: {e.result.triggered_rules}")
    if e.result.violations:
        print("  Invariant violations:")
        for v in e.result.violations:
            print(f"    - {v['proof']}/{v['invariant']}: {v['message']}")
    sys.exit(1)
except Exception as e:
    print(f"\nUnexpected error: {e}", file=sys.stderr)
    sys.exit(1)

status = handler.engine.status()
print(f"\n[engine] {handler.event_count} events evaluated")
